crates/proof-of-sql/sol_src/proof_primitive/sumcheck/Sumcheck.sol

```diff
@@ -0,0 +1,53 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.13;
+
+library Sumcheck {
+    error InvalidProofLength();
+    error RoundEvaluationMismatch();
+
+    struct Proof {
+        uint256[] coefficients;
+    }
+
+    struct Subclaim {
+        uint256[] evaluationPoint;
+        uint256 expectedEvaluation;
+    }
+
+    function verifyWithoutEvaluation(
+        Proof calldata proof,
+        bytes32 transcript,
+        uint64 degree,
+        uint64 numVariables,
+        uint256 claimedSum,
+        uint256 modulus
+    ) public pure returns (Subclaim memory result, bytes32 resultTranscript) {
+        if (proof.coefficients.length != numVariables * (degree + 1)) revert InvalidProofLength();
+        result.evaluationPoint = new uint256[](numVariables);
+        transcript = keccak256(abi.encodePacked(transcript, degree, numVariables));
+        for (uint64 roundIndex = 0; roundIndex < numVariables; ++roundIndex) {
+            uint256 startIndex = roundIndex * (degree + 1);
+            transcript = keccak256(abi.encodePacked(transcript, proof.coefficients[startIndex:startIndex + degree + 1]));
+            uint256 roundEvaluationPoint = uint256(transcript);
```

Comment:
Is this intended to be a custom transcript protocol?

In the merlin protocol, drawing a challenge would also update the state of the transcript.

Reply:
Here's the Rust implementation of the same thing:
https://github.com/spaceandtimelabs/sxt-proof-of-sql/blob/main/crates/proof-of-sql/src/base/proof/keccak256_transcript.rs

The intent here is that drawing a challenge always occurs after anything is appended to the transcript. That way, there is only one hash per challenge/response round. So, the current state of the transcript _is_ the current challenge.

If I need two challenges in a row, that would mean the prover should send an empty message and we would do: `transcript = keccak256(abi.encodePacked(transcript));`